flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java

```diff
@@ -293,20 +325,79 @@ protected InputChannel createKnownInputChannel(
     static SupplierWithException<BufferPool, IOException> createBufferPoolFactory(
-            BufferPoolFactory bufferPoolFactory, int floatingNetworkBuffersPerGate) {
-        Pair<Integer, Integer> pair =
-                NettyShuffleUtils.getMinMaxFloatingBuffersPerInputGate(
-                        floatingNetworkBuffersPerGate);
+            BufferPoolFactory bufferPoolFactory,
+            int minFloatingBuffersPerGate,
+            int maxFloatingBuffersPerGate) {
+        Pair<Integer, Integer> pair = Pair.of(minFloatingBuffersPerGate, maxFloatingBuffersPerGate);
         return () -> bufferPoolFactory.createBufferPool(pair.getLeft(), pair.getRight());
     }
 
+    /**
+     * Based on whether the used exclusive buffers exceed the threshold, decide whether all buffers
+     * in the gate use Floating Buffers.
+     *
+     * <p>The threshold is configured by {@link
+     * NettyShuffleEnvironmentOptions#NETWORK_READ_BUFFERS_REQUIRED_PER_GATE_MAX}. If the option is
+     * not configured, the threshold for Blocking jobs is {@link
+     * NettyShuffleUtils#DEFAULT_MAX_BUFFERS_PER_GATE_FOR_BATCH} and the threshold for Streaming
+     * jobs is {#link NettyShuffleUtils#DEFAULT_MAX_BUFFERS_PER_GATE_FOR_STREAMING}.
+     */
+    protected class GateBuffersNumCalculator {
+
+        private final int minFloatingBuffers;
+
+        private final int maxFloatingBuffers;
+
+        private final int actualExclusiveBuffersPerChannel;
+
+        GateBuffersNumCalculator(ResultPartitionType partitionType, int numInputChannels) {
+            int maxGateBuffersThreshold =
+                    maxRequiredBuffersPerGate(partitionType, requiredMaxBuffersPerGate);
+            int expectMinGateBuffers =
+                    getExpectMinBuffersPerGate(
+                            numInputChannels, configuredNetworkBuffersPerChannel);
+            int expectMaxGateBuffers =
+                    getExpectMaxBuffersPerGate(
+                            numInputChannels,
+                            configuredNetworkBuffersPerChannel,
+                            floatingNetworkBuffersPerGate);
+            int minGateBuffers = Math.min(maxGateBuffersThreshold, expectMinGateBuffers);
+
+            this.actualExclusiveBuffersPerChannel =
+                    adjustExclusiveBuffersPerChannel(
+                            configuredNetworkBuffersPerChannel,
+                            numInputChannels,
+                            maxGateBuffersThreshold);
+
+            int actualExclusiveBuffersPerGate =
+                    getActualExclusiveBuffersInGate(
+                            numInputChannels, actualExclusiveBuffersPerChannel);
+            this.minFloatingBuffers = Math.max(1, minGateBuffers - actualExclusiveBuffersPerGate);
+            this.maxFloatingBuffers =
+                    Math.max(1, expectMaxGateBuffers - actualExclusiveBuffersPerGate);
+            checkState(minFloatingBuffers <= maxFloatingBuffers);
+        }
```

Comment:
The calculation logics should not be in the constructor.

We can have a private constructor, and a public static factory method that performs the calculations.